### PR TITLE
feat: allow state extension

### PR DESCRIPTION
**Pull request by jleach**
# Summary of Changes

This PR is the first step in allowing developers to extend the default state as needed.

# Related Issues

Please reference here any issue #'s that are relevant to this PR, or simply enter "N/A" if this PR does not relate to any existing issues.

# Pull Request Checklist

Tick all boxes below to demonstrate that you have completed the respective task. If the item does not apply to your this PR **check it anyway** to make it apparent that there's nothing to do.

- [x] All commits contain a DCO `Signed-off-by` line (we use the [DCO GitHub app](https://github.com/apps/dco) to enforce this);
- [x] Updated LICENSE-3RD-PARTY.md for any added dependencies or vendored components;
- [x] Updated documentation as needed for changed code and new or modified features;
- [x] Added sufficient [tests](../__tests__/) so that overall code coverage is not reduced.

If you have _any_ questions to _any_ of the points above, just **submit and ask**! This checklist is here to _help_ you, not to deter you from contributing!

Pro Tip 🤓

- Read our [contribution guide](../CONTRIBUTING.md) at least once; it will save you a few review cycles!
- Your PR will likely not be reviewed until all the above boxes are checked and all automated tests have passed.

_PR template adapted from the Python attrs project._

### CODE REVIEW

**Comment by jleach** [comment]
@cvarjao @amanji I think this part is ready to merge. I've tested extending it and it seems to work fine. A little polishing perhaps on the Typescript types but I'm not sure what that looks like now. Only that when I merge the reducers I need to `// @ts-ignore` some ts complaints.

File: core/App/contexts/store.tsx [hunk 1]
```diff
@@ -1,53 +1,96 @@
...
+import _defaultReducer, { ReducerAction } from './reducers/store'
+
+type Reducer = <S extends State>(state: S, action: ReducerAction) => S
+
+interface StoreProviderProps {
+  initialState?: State
+  reducer?: Reducer
+}
+
+export class DefaultBifoldState implements State {
```
Comment (amanji) [inline core/App/contexts/store.tsx]: Im not sure if I agree with converting to classes. Interfaces don't actually get compiled which means smaller builds and there's no overhead of instantiating.

**Review by amanji — CHANGES_REQUESTED** [review]

File: core/App/index.ts [hunk 2]
```diff
@@ -55,6 +56,8 @@ export type { GenericFn } from './types/fn'
 export type { AuthenticateStackParams } from './types/navigators'
 export type { OnboardingStyleSheet }
 export type { WalletSecret } from './types/security'
+// export type { State as DefaultState } from './types/state'
```
Comment (amanji) [inline core/App/index.ts]: Dead code?

File: core/App/contexts/store.tsx (hunk 1 above)
Comment (jleach) [reply core/App/contexts/store.tsx]: Let me see if I can revert it without TS complaining too much.

File: core/App/contexts/store.tsx [hunk 3]
```diff
@@ -2,41 +2,49 @@ import React, { createContext, Dispatch, useContext, useReducer } from 'react'
...
+  },
+  credential: {
+    revoked: new Set(),
+    revokedMessageDismissed: new Set(),
+  },
+  error: null,
+  loading: false,
+}
+
+export const createInitialStateFactory = (): State => {
```
Comment (amanji) [inline core/App/contexts/store.tsx]: Why do we need this if we can just return `initialState`?

**Review by amanji — DISMISSED** [review]

**Review by amanji — APPROVED** [review]

**Pull request merged**